compiler/src/compiler/optimize/inlining.rs

```diff
@@ -0,0 +1,143 @@
+            return Err("Tried to inline, but the expression is not a call.");
+        };
+        let Expression::Lambda {
+            parameters,
+            responsible_parameter,
+            body,
+        } = visible.get(*function) else {
+            return Err("Tried to inline, but the call's receiver is not a lambda.");
+        };
+        if arguments.len() != parameters.len() {
+            return Err("Tried to inline, but the number of arguments doesn't match the expected parameter count.");
+        }
+
+        let id_mapping: HashMap<Id, Id> = parameters
+            .iter()
+            .zip(arguments.iter())
+            .map(|(parameter, argument)| (*parameter, *argument))
+            .chain([(*responsible_parameter, *responsible_argument)])
+            .chain(
+                body.defined_ids()
+                    .into_iter()
+                    .map(|id| (id, id_generator.generate())),
+            )
+            .collect();
+        let mut inlined_body = body.clone();
+        inlined_body.replace_ids(&mut |id| {
+            if let Some(replacement) = id_mapping.get(id) {
+                *id = *replacement;
+            }
+        });
+
+        *self = Expression::Multiple(inlined_body);
+
+        Ok(())
+    }
+}
+
+impl Mir {
+    pub fn inline_functions_containing_use(&mut self) {
+        let mut functions_with_use = HashSet::new();
+        for (id, expression) in self.body.iter() {
+            if let Expression::Lambda { body, .. } = expression &&
+                    body.iter().any(|(_, expr)| matches!(expr, Expression::UseModule { .. })) {
+                functions_with_use.insert(id);
+            }
+        }
+
+        self.body.visit_with_visible(&mut |_, expression, visible, _| {
+            if let Expression::Call { function, .. } = expression && functions_with_use.contains(function) {
+                // If inlining fails with an `Err`, there's nothing we can do
+                // except apply other optimizations first and then try again
+                // later.
+                let _ = expression.inline_call(visible, &mut self.id_generator);
+            }
+        });
+    }
+
+    pub fn inline_functions_of_maximum_complexity(&mut self, complexity: Complexity) {
+        let mut small_functions = HashSet::new();
+        for (id, expression) in self.body.iter() {
+            if let Expression::Lambda { body, .. } = expression && body.complexity() <= complexity {
+                small_functions.insert(id);
+            }
+        }
+
+        self.body.visit_with_visible(&mut |_, expression, visible, _| {
+            if let Expression::Call { function, .. } = expression && small_functions.contains(function) {
+                let _ = expression.inline_call(visible, &mut self.id_generator);
+            }
+        });
+    }
+
+    pub fn inline_functions_only_called_once(&mut self) {
+        let mut ids_to_number_of_references: HashMap<Id, usize> = HashMap::new();
```

Comment:
```suggestion
        let mut reference_counts: HashMap<Id, usize> = HashMap::new();
```

Shorter, but maybe not as clear